[HttpFoundation] Add a way to anonymize IPs

Seldaek opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features --> TODO

This is helpful for GDPR compliance reasons, and it isn't much code saved but it's also good if you don't have to think about how to do it.

Seldaek committed:
[HttpFoundation] Add a way to anonymize IPs

Seldaek commented on src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php:
```diff
@@ -101,4 +101,25 @@ public function invalidIpAddressData()
+            ['1.2.3.4', '1.2.3.0'],
+            ['2a01:198:603:0:396e:4789:8e99:890f', '2a01:198:603:0:396e:4789:8e99:0'],
+            ['0:0:0:0:0:0:0:1', '0:0:0:0:0:0:0:0'],
+            ['0:0:603:0:396e:4789:8e99:0001', '0:0:603:0:396e:4789:8e99:0'],
```
For IPv6, maybe we should remove more bits, I am not sure what would be considered anonymous enough. If someone feels like researching what would be appropriate be my guest :)

xabbuh commented on src/Symfony/Component/HttpFoundation/IpUtils.php:
```diff
@@ -153,4 +153,12 @@ public static function checkIp6($requestIp, $ip)
     }
+
+    /**
+     * Anonymizes an IP/IPv6 by removing its last bit
```
Shouldn't "bit" be "byte" here?

Seldaek replied:
Maybe, would like to first figure out what to do with IPv6s

javiereguiluz replied:
Maybe we could use an alternative anonymizing method based on IP masks:

```php
public static function anonymize(string $ip): string
{
    $packedAddress = inet_pton($ip);
    $anonymizingMask = 4 === strlen($packedAddress) ? '255.255.255.0' : 'ffff:ffff:ffff:ffff:0000:0000:0000:0000';

    return inet_ntop($packedAddress & inet_pton($anonymizingMask));
}
```

Code adapted from https://github.com/geertw/php-ip-anonymizer/blob/master/src/IpAnonymizer.php (MIT license)

stof replied:
this comment is still true today. Even for IPv4, it removes a byte, not a bit

nicolas-grekas replied:
I think so :)

stof commented on src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php:
```diff
@@ -101,4 +101,25 @@ public function invalidIpAddressData()
+        return [
+            ['192.168.1.1', '192.168.1.0'],
+            ['1.2.3.4', '1.2.3.0'],
+            ['2a01:198:603:0:396e:4789:8e99:890f', '2a01:198:603:0:396e:4789:8e99:0'],
```
more cases should be tested, especially regarding ipv6. See https://github.com/geertw/php-ip-anonymizer/pull/3 for some cases used there.

Seldaek commented:
Ok I updated wording/docblock, added tests and support for IPv4-mapped IPv6s (which complicates things a little, if anyone sees a better way to do this feel free, as long as the tests pass..), and now remove the last 8 bytes (/64 equivalent) of IPv6s.

For me it's ready to ship now but any feedback is appreciated.

fabpot commented:
Thank you @Seldaek.
